### PR TITLE
New remote control modules (LIGHT, AUDIO, HMI_SETTINGS) and parameters (SIS Data)

### PR DESCRIPTION
**Pull request by yang1070**
SDL remote control provides the ability for mobile applications to control certain components of the vehicle, such as, tune the radio, set the desired temperature or change the fan speed. However, currently only `CLIMATE` and `RADIO` modules are available for remote-control. Mobile application developers need more than basic climate or radio control. To meet developers' requirements, we propose to add three new modules (`AUDIO`, `LIGHT`, `HMI_SETTINGS`) and some new parameters (heatedXXXEnable and radio sisData) to the existing `CLIMMATE` and `RADIO` modules respectively. This proposal only defines new data types and new parameters added to the existing data types. It does not add or change any remote control RPCs.

### CODE REVIEW

**Comment by yang1070** [comment]
@theresalech This is ready for review. Thanks.

File: proposals/NNNN-new-remote-control-modules-and-parameters.md [hunk 1]
```diff
@@ -0,0 +1,512 @@
+# New remote control modules (LIGHT, ADUIO, HMI_SETTINGS) and parameters (SIS Data)
+
+* Proposal: [SDL-NNNN](NNNN-new-remote-control-modules-and-parameters.md)
+* Author: [Zhimin Yang](https://github.com/yang1070)
+* Status: **Awaiting review**
+* Impacted Platforms: [Core / iOS / Android / RPC ]
+
+## Introduction
+
+SDL remote control provides the ability for mobile applications to control certain components of the vehicle, such as, tune the radio, set the desired temperature or change the fan speed. However, currently only `CLIMATE` and `RADIO` modules are available for remote-control. Mobile application developers need more than basic climate or radio control. To meet developers' requirements, we propose to add three new modules (`AUDIO`, `LIGHT`, `HMI_SETTINGS`) and some new parameters (heatedXXXEnable and radio sisData) to the existing `CLIMMATE` and `RADIO` modules respectively. This proposal only defines new data types and new parameters added to the existing data types. It does not add or change any remote control RPCs.
```
Comment (theresalech) [inline proposals/NNNN-new-remote-control-modules-and-parameters.md]: Please fix `CLIMATE` here and in the PR description.

File: proposals/NNNN-new-remote-control-modules-and-parameters.md [hunk 2]
```diff
@@ -0,0 +1,512 @@
...
+
+## Motivation
+
+Mobile application developers would like to develop applications that take fully advantage of the capability that the vehicle provides. There are many cases that mobile application developers ask for new functions of SDL remote control. For example,
```
Comment (theresalech) [inline proposals/NNNN-new-remote-control-modules-and-parameters.md]: Please update to "full advantage"

File: proposals/NNNN-new-remote-control-modules-and-parameters.md [hunk 3]
```diff
@@ -0,0 +1,512 @@
...
+- Some light control applications want to change the ambient lights of the vehicle and/or turn on/off certain exterior lights (while the vehicle is in parking).
+
+- Some radio control applications need more than the traditional RDS data, they need HD radio status data (especially the station id in the SIS data).
+
+
+## Proposed solution
+
+As usual, the remote control mobile applications use the following RPCs
+- `GetSystemCapability` with `SystemCapabilityType=REMOTE_CONTROL` to read available control modules and available control items within each module;
+- `GetInteriorVechileData` to read the current status value of a remote-control module;
```
Comment (theresalech) [inline proposals/NNNN-new-remote-control-modules-and-parameters.md]: I believe this should be `GetInteriorVehicleData`

File: proposals/NNNN-new-remote-control-modules-and-parameters.md [hunk 4]
```diff
@@ -0,0 +1,512 @@
...
+- `GetInteriorVechileData` with `subscribe`=true/false to register/un-register a module's data change notifications;
```
Comment (theresalech) [inline proposals/NNNN-new-remote-control-modules-and-parameters.md]: I believe this should be `GetInteriorVehicleData`

File: proposals/NNNN-new-remote-control-modules-and-parameters.md [hunk 5]
```diff
@@ -0,0 +1,512 @@
...
+- `SetInteriorVechileData` to change the settings/status of a remote-control module;
```
Comment (theresalech) [inline proposals/NNNN-new-remote-control-modules-and-parameters.md]: I believe this should be `SetInteriorVehicleData`

File: proposals/NNNN-new-remote-control-modules-and-parameters.md [hunk 6]
```diff
@@ -0,0 +1,512 @@
...
+- `OnInteriorVechileData` to receive any remote-control module value change notifications;
```
Comment (theresalech) [inline proposals/NNNN-new-remote-control-modules-and-parameters.md]: I believe this should be `OnInteriorVehicleData`

File: proposals/NNNN-new-remote-control-modules-and-parameters.md [hunk 7]
```diff
@@ -0,0 +1,512 @@
...
+
+In addition, in a GetInteriorVechileData response or OnInteriorVechileData notifications, an `xyzmoduleControlData` indicates the current value of the module status; in SetInteriorVechileData request, an `xyzmoduleControlData` indicates the required target state of the module.
```
Comment (theresalech) [inline proposals/NNNN-new-remote-control-modules-and-parameters.md]: I believe this should be `GetInteriorVehicleData`, `OnInteriorVehicleData`, and "SetInteriorVehicleData`

File: proposals/NNNN-new-remote-control-modules-and-parameters.md [hunk 8]
```diff
@@ -0,0 +1,512 @@
...
+
+#### Add there new module types to the existing module list.
```
Comment (theresalech) [inline proposals/NNNN-new-remote-control-modules-and-parameters.md]: Please change to "Add new module types to the existing module list."

File: proposals/NNNN-new-remote-control-modules-and-parameters.md [hunk 9]
```diff
@@ -0,0 +1,512 @@
...
+| Audio Source | MOBILE_APP, RADIO_TUNER, CD, BLUETOOTH, USB, etc. see PrimaryAudioSource| Get/Set/Notification | defines one of the available audio sources |
+| keep Context | true, false| Set only | control whether HMI shall keep current application context or switch to default media UI/APP associated with the audio source|
+| Equilizer Settings | Struct {Channel ID as integer, Channel setting as 0%-100%} | Get/Set/Notification | Defines the list of supported channels (band) and their current/desired settings on HMI
+
+Any remote control applications can read the current audio data. A RC and media application can set audio source to either itself (with target `MOBILE_APP`) or other system sources. It CANNOT set audio source to other mobile applications. When a mobile application sends a SetInteriorVechileData request to change the audio source from MOBILE_APP to other types of audio source without `keepContext` parameter or with `keepContext=false`, after the system successfully executes the request, the application will go to HMI level 'BACKGROUND'. If in the same request `keepContext`=true, the system shall not change this application's HMI level.
```
Comment (theresalech) [inline proposals/NNNN-new-remote-control-modules-and-parameters.md]: Please update to `SetInteriorVehicleData`

File: proposals/NNNN-new-remote-control-modules-and-parameters.md [hunk 10]
```diff
@@ -0,0 +1,512 @@
...
+
+We recommend that SDL does not block audio source changing requests depending on HMI level. However, we recommend that HMI disallow applications running in background change the audio source in order to (1) prevent the application from stealing the audio, and (2) reduce confusion to the driver.  This restriction means (media and remote control type) applications running in HMI level `FULL` or `LIMTED` (already has access to audio) can change the audio source. Applications in HMI level `NONE` or `BACKGROUND` are not allowed to change the audio source. Instead, if an application running in `BACKGROUND` wants to switch the audio source from others to itself, the application shall send an `alert` with at least "yes" and "no" soft buttons to notify the driver the intention to change the audio source, and set value `STEAL_FOCUS` (the system shall bring the application to foreground and set audio source to this media application) as the `SystemAction` of the "yes" soft button, so that the driver can click the soft button to confirm the switch.
```
Comment (theresalech) [inline proposals/NNNN-new-remote-control-modules-and-parameters.md]: Please update to `LIMITED`

File: proposals/NNNN-new-remote-control-modules-and-parameters.md [hunk 11]
```diff
@@ -0,0 +1,512 @@
...
+
+If the target audio source is `RADIO_TUNER`, the system shall turn on the radio and start playing with the last used band (AM|FM|XM) and frequency or station. For an example, if a media and remote-control application running in `FULL` want to switch the audio source from itself to radio and tune to a specified band and frequency (not last used band and frequency), it shall do two SetInteriorVechileData RPC calls. The first one has `ADUIO` as targeted module type, `RADIO_TUNNER` as the new audio `source` and `keepContext`=true. The second one has `RADIO` as targeted module type, and include `band`=`AM|FM|XM` and corresponding desired frequency parameters.
```
Comment (theresalech) [inline proposals/NNNN-new-remote-control-modules-and-parameters.md]: Please update to `SetInteriorVehicleData` and `RADIO_TUNER`

File: proposals/NNNN-new-remote-control-modules-and-parameters.md [hunk 12]
```diff
@@ -0,0 +1,512 @@
...
+
+#### For "HMI Setting" remote-control mobile application will be able to READ and SET:
+| Control Item | Value Range |Type | Comments |
+| ------------ |------------ | ------------ | ------------ |
+| Display Mode | DAY, NIGHT, AUTO | Get/Set/Notification | Current display mode of the HMI display |
+| Distance Unit | MILES, KILOMETERS | Get/Set/Notification | Distance Unit used in the HMI (for maps/tracking distances) |
+| Temperature Unit | FAHRENHEIT, CELSIUS | Get/Set/Notification | Temperature Unit used in the HMI (for temperature measuring systems) |
+
+`AUTO` display mode is set only. The driver can set the display mode as `DAY`, `NIGHT` or `AUTO` via HMI or a mobile application. However, the current mode used by the HMI is either `DAY` or `NIGHT`. GetInteriorVechileData of HMI_SETTINGS module type shall return the current value of the display mode used in HMI, not the current settings.
```
Comment (theresalech) [inline proposals/NNNN-new-remote-control-modules-and-parameters.md]: `GetInteriorVehicleData`

File: proposals/NNNN-new-remote-control-modules-and-parameters.md [hunk 13]
```diff
@@ -0,0 +1,512 @@
...
+| Control Item | Value Range |Type | Comments |
+| ------------ |------------ | ------------ | ------------ |
+| HD radio SIS data | see sisData type| Get,Notification | read only HD radio SIS data |
+
+Like Radio Data System (RDS) is a communications protocol standard for embedding small amounts of digital information in conventional FM radio broadcasts, Station Information Service (SIS) data (described in NRSC-5-C and NRSC-4-B and http://www.nrscstandards.org/documentArchive/NRSC-5-A%202005/1020sEb.pdf) provides basic information for HD radio about the station such as call sign, as well as information not displayable to the consumer such as the station identification number.
+
+#### In addition to existing "CLIMATE" control items, remote-control mobile applications will be able to READ and SET :
+| Control Item | Value Range |Type | Comments |
+| ------------ |------------ | ------------ | ------------ |
+| heated windsheild | true, false| Get/Set/Notification | true means ON, false means OFF |
```
Comment (theresalech) [inline proposals/NNNN-new-remote-control-modules-and-parameters.md]: Please update to "windshield"

File: proposals/NNNN-new-remote-control-modules-and-parameters.md [hunk 14]
```diff
@@ -0,0 +1,512 @@
...
+
+### Mobile API XML changes
+The changes are listed below.
+```xml
+  <enum name="ModuleType">
+    :
+    :
+    <element name="AUDIO"/>
+    <element name="LIGHT"/>
+    <element name="HMI_SETTIGNS"/>
```
Comment (theresalech) [inline proposals/NNNN-new-remote-control-modules-and-parameters.md]: "HMI_SETTINGS"

File: proposals/NNNN-new-remote-control-modules-and-parameters.md [hunk 15]
```diff
@@ -0,0 +1,512 @@
...
+    <element name="USB2">
+    </element>
+    <element name="LINE_IN">
+    </element>
+    <element name="IPOD">
+    </element>
+    <element name="MOBILE_APP">
+    </element>
+    <!--new addition--->
+    <element name="RADIO_TUNNER">
```
Comment (theresalech) [inline proposals/NNNN-new-remote-control-modules-and-parameters.md]: "RADIO_TUNER"

File: proposals/NNNN-new-remote-control-modules-and-parameters.md [hunk 16]
```diff
@@ -0,0 +1,512 @@
...
+        True: Available, False: Not Available, Not present: Not Available.
+      </description>
+    </param>      
+  </struct>  
+
+  <struct name="ClimateControlData">
+    :
+    :
+    <param name="heatedSteeringWheelEnable" type="Boolean" mandatory="false">
+      <description>value false means disabled/turn off, value true mean enabled/turn on.</description>
```
Comment (theresalech) [inline proposals/NNNN-new-remote-control-modules-and-parameters.md]: Please update to "value true means"

File: proposals/NNNN-new-remote-control-modules-and-parameters.md [hunk 17]
```diff
@@ -0,0 +1,512 @@
...
+
+
+## Potential downsides
+
+None
+
+## Impact on existing code
+
+RPC changes:
+- Add new AUDIO, LIGHT,HIM_SETTINGS module and corresponding capabilities and control data types to RC APIs.
```
Comment (theresalech) [inline proposals/NNNN-new-remote-control-modules-and-parameters.md]: Please update to "LIGHT, HMI_SETTINGS"

File: proposals/NNNN-new-remote-control-modules-and-parameters.md [hunk 18]
```diff
@@ -0,0 +1,512 @@
...
+- The proposal has the impact on: SetInteriorVehicleData, GetInteriorVehicleData, OnInteriorVehicleData, GeSystemCapability
```
Comment (theresalech) [inline proposals/NNNN-new-remote-control-modules-and-parameters.md]: Please update to "GetSystemCapability"

File: proposals/NNNN-new-remote-control-modules-and-parameters.md (hunk 1 above)
Comment (yang1070) [reply proposals/NNNN-new-remote-control-modules-and-parameters.md]: thanks, updated CLIMMATE to CLIMATE

File: proposals/NNNN-new-remote-control-modules-and-parameters.md (hunk 2 above)
Comment (yang1070) [reply proposals/NNNN-new-remote-control-modules-and-parameters.md]: updated

File: proposals/NNNN-new-remote-control-modules-and-parameters.md (hunk 4 above)
Comment (yang1070) [reply proposals/NNNN-new-remote-control-modules-and-parameters.md]: thank you

File: proposals/NNNN-new-remote-control-modules-and-parameters.md (hunk 5 above)
Comment (yang1070) [reply proposals/NNNN-new-remote-control-modules-and-parameters.md]: thank you

File: proposals/NNNN-new-remote-control-modules-and-parameters.md (hunk 6 above)
Comment (yang1070) [reply proposals/NNNN-new-remote-control-modules-and-parameters.md]: updated

File: proposals/NNNN-new-remote-control-modules-and-parameters.md (hunk 7 above)
Comment (yang1070) [reply proposals/NNNN-new-remote-control-modules-and-parameters.md]: updated

File: proposals/NNNN-new-remote-control-modules-and-parameters.md (hunk 8 above)
Comment (yang1070) [reply proposals/NNNN-new-remote-control-modules-and-parameters.md]: there -> three

File: proposals/NNNN-new-remote-control-modules-and-parameters.md (hunk 9 above)
Comment (yang1070) [reply proposals/NNNN-new-remote-control-modules-and-parameters.md]: changed

File: proposals/NNNN-new-remote-control-modules-and-parameters.md (hunk 10 above)
Comment (yang1070) [reply proposals/NNNN-new-remote-control-modules-and-parameters.md]: changed

**Comment by theresalech** [comment]
@yang1070 Sorry I missed one typo - can you please update `AUDIO` in the title of the PR?

File: proposals/NNNN-new-remote-control-modules-and-parameters.md (hunk 11 above)
Comment (yang1070) [reply proposals/NNNN-new-remote-control-modules-and-parameters.md]: changed

File: proposals/NNNN-new-remote-control-modules-and-parameters.md (hunk 12 above)
Comment (yang1070) [reply proposals/NNNN-new-remote-control-modules-and-parameters.md]: changed

File: proposals/NNNN-new-remote-control-modules-and-parameters.md (hunk 13 above)
Comment (yang1070) [reply proposals/NNNN-new-remote-control-modules-and-parameters.md]: changed

File: proposals/NNNN-new-remote-control-modules-and-parameters.md (hunk 14 above)
Comment (yang1070) [reply proposals/NNNN-new-remote-control-modules-and-parameters.md]: changed

File: proposals/NNNN-new-remote-control-modules-and-parameters.md (hunk 15 above)
Comment (yang1070) [reply proposals/NNNN-new-remote-control-modules-and-parameters.md]: changed

File: proposals/NNNN-new-remote-control-modules-and-parameters.md (hunk 16 above)
Comment (yang1070) [reply proposals/NNNN-new-remote-control-modules-and-parameters.md]: changed

File: proposals/NNNN-new-remote-control-modules-and-parameters.md (hunk 17 above)
Comment (yang1070) [reply proposals/NNNN-new-remote-control-modules-and-parameters.md]: changed

File: proposals/NNNN-new-remote-control-modules-and-parameters.md (hunk 18 above)
Comment (yang1070) [reply proposals/NNNN-new-remote-control-modules-and-parameters.md]: changed

**Comment by yang1070** [comment]
@theresalech updated. Than you.

**Pull request merged**